### PR TITLE
Added new testing documentation

### PR DESCRIPTION
**Pull request by dereknola**
Signed-off-by: dereknola <derek.nola@suse.com>

#### Proposed Changes ####
New documentation that describes when and how to write new k3s tests.

#### Types of Changes ####
Documentation

#### Verification ####
N/A

#### Linked Issues ####
https://github.com/k3s-io/k3s/issues/3729

#### User-Facing Change ####

```release-note
NONE
```

### CODE REVIEW

File: tests/TESTING.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+___
+
+## Unit Tests
+
+Unit tests should be written when a component or function of a package needs testing.
+Unit tests should be used for "white box" testing.
+
+### Framework
+
+All unit tests in K3s follow a [Table Driven Test](https://github.com/golang/go/wiki/TableDrivenTests) style. Specifically, K3s unit tests are automatically generated using the [gotests](https://github.com/cweill/gotests) tool. This is built into the Golang vscode extension or can be run via command line. Additionally, a set of custom templates is provided to the tool to extend the generated test's functionality. To use these templates, call:
```
Comment (briandowns) [inline tests/TESTING.md:17]:
```suggestion
All unit tests in K3s follow a [Table Driven Test](https://github.com/golang/go/wiki/TableDrivenTests) style. Specifically, K3s unit tests are automatically generated using the [gotests](https://github.com/cweill/gotests) tool. This is built into the Go vscode extension or can be run via command line. Additionally, a set of custom templates is provided to the tool to extend the generated test's functionality. To use these templates, call:
```

File: tests/TESTING.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+Generally, E2E tests are run as a nightly Jenkins job for QA. They can still be run locally but additional setup may be required.
+
+```bash
+go test ./tests/... -run E2E
+```
+
+## Contributing New Tests
+
+___
+We gladly accept new tests of all types. If you wish to create
```
Comment (briandowns) [inline tests/TESTING.md:102]:
```suggestion
We gladly accept new and updated tests of all types. If you wish to create
```

Since we're going for table tests for unit tests, I'd like to encourage others to add new entries that may have been missed that cover previously untested code paths or account for edge/corner cases.

File: tests/TESTING.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+a new test, please submit a PR with a title that includes the words `<NAME_OF_TEST> Created`. Additionally, the final merge commit into K3s should also include those keywords. This
```
Comment (briandowns) [inline tests/TESTING.md:103]:
```suggestion
a new test or update an existing test, please submit a PR with a title that includes the words `<NAME_OF_TEST> Created`. Additionally, the final merge commit into K3s should also include those keywords. This
```

File: tests/TESTING.md [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+### Running
+
+Generally, E2E tests are run as a nightly Jenkins job for QA. They can still be run locally but additional setup may be required.
+
+```bash
+go test ./tests/... -run E2E
+```
+
+## Contributing New Tests
```
Comment (briandowns) [inline tests/TESTING.md:99]:
```suggestion
## Contributing New Tests or Updated Tests
```

File: tests/TESTING.md [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+```
+
+Note that as unit tests call functions directly, they are the primary drivers of K3s's code coverage
+metric.
+
+___
+
+## Integration Tests
+
+Integration tests should be used to test a specific functionality of k3s that exists across multiple go packages, either via exported function calls, or more often, CLI comands.
```
Comment (briandowns) [inline tests/TESTING.md:48]:
```suggestion
Integration tests should be used to test a specific functionality of k3s that exists across multiple Go packages, either via exported function calls, or more often, CLI comands.
```

File: tests/TESTING.md [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+go test ./tests/... -run E2E
+```
+
+## Contributing New Tests
+
+___
+We gladly accept new tests of all types. If you wish to create
+a new test, please submit a PR with a title that includes the words `<NAME_OF_TEST> Created`. Additionally, the final merge commit into K3s should also include those keywords. This
+helps us track and identify which release a test was first created. If a test needs to
+be backported into a previous release, the commit description should note that.
```
[the remark this answers is not in the document]
Comment (dereknola) [reply tests/TESTING.md:105]: Cut out the last 2 sentences. Community members don't need to know why, and core team members should already be aware of it. PRs also now automatically contain a backport area.

File: tests/TESTING.md (hunk 5 above)
Comment (dereknola) [reply tests/TESTING.md]: Fix

File: tests/TESTING.md (hunk 4 above)
Comment (dereknola) [reply tests/TESTING.md]: Fixed

File: tests/TESTING.md (hunk 3 above)
Comment (dereknola) [reply tests/TESTING.md]: Fixed

File: tests/TESTING.md (hunk 2 above)
Comment (dereknola) [reply tests/TESTING.md]: Fixed

File: tests/TESTING.md (hunk 1 above)
Comment (dereknola) [reply tests/TESTING.md]: Fixed

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -10,6 +10,8 @@ If you're interested in contributing documentation, please note the following:
 - Doc issues are raised in this repository, and they are tracked under the `kind/documentation` label.
 - Pull requests are submitted to the K3s documentation source in the [Rancher docs repository.](https://github.com/rancher/docs/) The K3s docs content is in the `content/k3s/` directory.
 
+If you're intested in contributing new tests, please see the `TESTING.md` in the tests directory.
```
[the remark this answers is not in the document]
Comment (dereknola) [reply CONTRIBUTING.md]: Fixed

**Review by briandowns — APPROVED** [review]

**Review by brandond — APPROVED** [review]
